### PR TITLE
safe_fork: unblock most signals before waiting for child

### PR DESCRIPTION
**Pull request by floppym**
This ensures we will recieve SIGTSTP if the user presses Ctrl-Z.

We continue blocking SIGCHLD to ensure the child is processed by
wait_for_terminate_and_check.

Fixes: https://github.com/systemd/systemd/issues/9806

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
Yep, works for me!

**Comment by floppym** [comment]
I also notice that we don't unblock `SIGCHLD` if there is an error returned from `wait_for_terminate_and_check`. We probably need to restore the original signal mask before any return statement is executed. Maybe this deserves a cleanup function?

**Comment by floppym** [comment]
The failing test is some kind of timeout. I can't really understand what is happening from the log, but I doubt this PR is the cause.

**Comment by floppym** [comment]
@poettering @keszybz @yuwata Could someone take a look at this PR? Would be nice if it could be included in v245.

**Comment by keszybz** [comment]
Oh, I missed this one. It'll have to go in post-release.

**Comment by floppym** [comment]
Ok. I just didn't want this to get lost.

File: src/basic/process-util.c [hunk 1]
```diff
@@ -1178,6 +1178,14 @@ int must_be_root(void) {
         return log_error_errno(SYNTHETIC_ERRNO(EPERM), "Need to be root.");
 }
 
+static void restore_sigsetp(sigset_t **ssp)
+{
```
Comment (keszybz) [inline src/basic/process-util.c]: { on previous line please.

File: src/basic/process-util.c [hunk 2]
```diff
@@ -1178,6 +1178,14 @@ int must_be_root(void) {
...
+        if (*ssp) {
+                (void) sigprocmask(SIG_SETMASK, *ssp, NULL);
+                *ssp = NULL;
```
Comment (keszybz) [inline src/basic/process-util.c]: Assigning to `*ssp` is not necesary...

**Pull request merged**